### PR TITLE
pydantic v2 migration

### PR DESCRIPTION
**Pull request by kashif**
Initial pydantic v2 migraton

### CODE REVIEW

File: requirements/requirements.txt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 numpy~=1.16
 pandas>=1.0,<3
-pydantic~=1.7
+pydantic~=2.3
+pydantic-settings
```
Comment (ddelange) [inline requirements/requirements.txt]: what is `pydantic-settings` needed for? probably also needs a lower version pin

File: src/gluonts/shell/serve/__init__.py [hunk 2]
```diff
@@ -17,7 +17,7 @@
 from typing import List, Optional, Type, Union
 
 from flask import Flask
-from pydantic import BaseSettings
+from pydantic_settings import BaseSettings, SettingsConfigDict
```
Comment (kashif) [inline src/gluonts/shell/serve/__init__.py:20]: @ddelange its needed for the `BaseSettings` class

File: requirements/requirements.txt (hunk 1 above)
Comment (ddelange) [reply requirements/requirements.txt]:
```suggestion
pydantic-settings>=2.0.3
```

**Comment by kashif** [comment]
@ddelange  can you help me fix up the tests?

File: requirements/requirements.txt [hunk 3]
```diff
@@ -1,6 +1,7 @@
 numpy~=1.16
 pandas>=1.0,<3
-pydantic~=1.7
+pydantic~=2.3
```
Comment (ddelange) [inline requirements/requirements.txt]: Could we have:
```suggestion
pydantic~=2.1
```
in order to support pytorch lightning which is required by autogluon?


```
gluonts 0.13.8+gdd9383d3a1 depends on pydantic~=2.3
pydantic-settings 2.0.3 depends on pydantic>=2.0.1
fastapi 0.103.2 depends on pydantic!=1.8, !=1.8.1, !=2.0.0, !=2.0.1, !=2.1.0, <3.0.0 and >=1.7.4
lightning 2.0.9.post0 depends on pydantic<2.2.0 and >=1.7.4
```

**Comment by ddelange** [comment]
```
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/gluonts/ext/rotbaum/_types.py:28: in FeatureImportanceResult
    @model_validator()
E   TypeError: model_validator() missing 1 required keyword-only argument: 'mode'
```

```
mode: Specifies whether to validate the fields before or after validation.
```

https://docs.pydantic.dev/latest/api/functional_validators

**Comment by ddelange** [comment]
```
test/core/test_serde_dataclass.py:22: in <module>
    class Estimator:
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/gluonts/core/serde/_dataclass.py:139: in dataclass
    return _dataclass(cls, init, repr, eq, order, unsafe_hash, frozen)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/gluonts/core/serde/_dataclass.py:155: in _dataclass
    pydantic.dataclasses.dataclass(
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pydantic/dataclasses.py:141: in dataclass
    assert init is False, 'pydantic.dataclasses.dataclass only supports init=False'
E   AssertionError: pydantic.dataclasses.dataclass only supports init=False
```
https://github.com/awslabs/gluonts/actions/runs/6611474968/job/17955418303?pr=3007#step:5:75

**Comment by ddelange** [comment]
```
E       pydantic_core._pydantic_core.ValidationError: 1 validation error for CategoricalFeatureInfo
E       cardinality
E         Input should be a valid string [type=string_type, input_value=10, input_type=int]
E           For further information visit https://errors.pydantic.dev/2.4/v/string_type
```
https://github.com/awslabs/gluonts/actions/runs/6611474968/job/17955418303?pr=3007#step:5:701

**Comment by ddelange** [comment]
Hi @kashif 👋

I'm trying a less intrusive approach in https://github.com/awslabs/gluonts/pull/3026

**Pull request closed without merging**